File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.util.Scanner;

public class Main {
    static int[] board;
    static int n;
    static int cnt = 0;
    
    private static boolean check(int row) {
        for (int  i = 0; i < row; i++) {
            if (board[i] == board[row]) return false;
            if (Math.abs(row - i) == Math.abs(board[row] - board[i])) return false;
        }
        
        return true;
    }

    private static void backtracking(int row) {
        if (row == n) {
            cnt++;
            return;
        }
        
        for (int i = 0; i < n; i++) {
            board[row] = i;
            if (check(row)) {
                backtracking(row + 1);
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        board = new int[n];
        backtracking(0);
        System.out.println(cnt);
    }
}

File: 백준/Gold/9663. N－Queen/N－Queen.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import kotlin.math.abs

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()
    val board = IntArray(n)
    var count = 0

    fun isValid(row: Int): Boolean {
        return (0 until row).all {
            board[it] != board[row] && abs(row - it) != abs(board[row] - board[it])
        }
    }

    fun backtracking(row: Int) {
        if (row == n) {
            count++
            return
        }

        for (col in 0 until n) {
            board[row] = col
            if (isValid(row)) backtracking(row + 1)
        }
    }

    backtracking(0)
    println(count)
}